src/web/app/pages-session/session-submission-page/session-submission-page.component.ts

```diff
@@ -955,4 +997,120 @@ export class SessionSubmissionPageComponent implements OnInit, AfterViewInit {
+    } else if (selectedView === SessionView.GROUP_RECIPIENTS) {
+      this.currentSelectedSessionView = SessionView.GROUP_RECIPIENTS;
+      this.groupQuestionsByRecipient();
+    }
+  }
+
+  /**
+   * Group questions by recipients in {@code GROUP_RECIPIENTS} view.
+   */
+  groupQuestionsByRecipient(): void {
+    if (!this.hasLoadedAllRecipients) {
+      // We first need to load the recipient for all the questions. This is because questions with
+      // FIXED_RECIPIENT question submission mode are ungroupable and to know whether the question
+      // submission mode of a question, we need to load the recipient list first.
+      const recipientsObservables: Observable<FeedbackQuestionRecipients>[] = [];
+      const questionsToBeLoaded: QuestionSubmissionFormModel[] = [];
+
+      this.questionSubmissionForms.forEach((model: QuestionSubmissionFormModel) => {
+        if (!model.isLoading && !model.isLoaded) {
+          questionsToBeLoaded.push(model);
+          recipientsObservables.push(this.feedbackQuestionsService.loadFeedbackQuestionRecipients({
+            questionId: model.feedbackQuestionId,
+            intent: this.intent,
+            key: this.regKey,
+            moderatedPerson: this.moderatedPerson,
+            previewAs: this.previewAsPerson,
+          }));
+        }
+      });
+
+      // Find the groupable and ungroupable questions and construct the recipient to question mapping.
+      forkJoin(recipientsObservables)
+          .pipe(finalize(() => {
+            this.ungroupableQuestionsSorted = Array.from(this.ungroupableQuestions).sort();
+            this.hasLoadedAllRecipients = true;
+          }))
+          .subscribe({
+            next: (recipients: FeedbackQuestionRecipients[]) => {
+                for (let i = 0; i < recipients.length; i += 1) {
+                  const question: QuestionSubmissionFormModel = questionsToBeLoaded[i];
+                  // Only questions with question submission form mode being FIXED_RECIPIENT and with question type
+                  // not being CONSTSUM_RECIPIENTS, RANK_RECIPIENTS, and CONTRIB, are the groupable questions.
+                  if (this.getQuestionSubmissionFormMode(question, recipients[i].recipients.length)
+                      === QuestionSubmissionFormMode.FIXED_RECIPIENT
+                      && question.questionType !== FeedbackQuestionType.CONSTSUM_RECIPIENTS
+                      && question.questionType !== FeedbackQuestionType.RANK_RECIPIENTS
+                      && question.questionType !== FeedbackQuestionType.CONTRIB) {
+
+                    for (let j = 0; j < recipients[i].recipients.length; j += 1) {
+                      const recipient: FeedbackQuestionRecipient = recipients[i].recipients[j];
+                      this.addQuestionForRecipient(recipient.identifier, question.questionNumber);
+                    }
+                  } else {
+                    this.ungroupableQuestions.add(question.questionNumber);
```

Comment:
Might be safer to initialise the question arrays to empty before the fork join. If there are any bugs, the array could be left half with some values. This could be a potential issue if retry is implemented.

Reply:
Items are also added in the `loadFeedbackQuestionRecipientsForQuestion` function at line 572 (or somewhere around these lines), so we cannot set it to empty array right before forkjoin. Items are added in  `loadFeedbackQuestionRecipientsForQuestion` function is because as we scroll down the page, the feedback questions will be loaded when it is in view port, and when the feedback questions are loaded, the recipients are also loaded. So this is to prevent redundant http call to the backend.

Reply:
thanks for checking :)

Hopefully with v9 we are able to reduce some of this complexity. The frontend pages are extremely time consuming to review at the moment.